src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs

```diff
@@ -336,12 +352,16 @@ private static bool TryGetClassDeclarationList(INamedTypeSymbol typeSymbol, [Not
                 CollectionType collectionType = typeGenerationSpec.CollectionType;
                 switch (generationMode)
                 {
-                    case JsonSourceGenerationMode.MetadataAndSerialization:
+                    case JsonSourceGenerationMode.Metadata | JsonSourceGenerationMode.Serialization:
+                        typeGenerationSpec.GenerateSerializationLogic = typeGenerationSpec.FastPathIsSupported();
```

Comment:
Why do we need to set `GenerateSerializationLogic` here? If it is left unset, it gets defaulted to `FastPathIsSupported()` in TypeGenerationSpec.cs

Reply:
The [defaulting logic](https://github.com/dotnet/runtime/blob/e939506d2178947bfdcc4abba3d46812c786d3c0/src/libraries/System.Text.Json/gen/TypeGenerationSpec.cs#L39) determines `GenerateSerializationLogic`'s value based on the global setting `JsonSourceGenerationOptions.GenerationMode`. Setting it manually here allows the type-specific option on `JsonSerializableAttribute.Generation` mode can win over the globl setting.

Reply:
I don't understand.

The `generationMode` is getting flown into `TypeGenerationSpec.Initialize`. So the `TypeGenerationSpec` should have the `generationMode` set on it.

But I see that `TypeGenerationSpec`s are re-used, so you could have created one earlier for a property of a different type, and now are seeing a `JsonSerializableAttribute` for that type again, specifying a different mode.

Thinking more about this, I think we have a bug here:

```C#
    [JsonSerializable(typeof(Foo), GenerationMode = JsonSourceGenerationMode.Metadata)]
    internal partial class Context1 : JsonSerializerContext
    {
    }

    [JsonSerializable(typeof(Foo), GenerationMode = JsonSourceGenerationMode.Serialization)]
    internal partial class Context2 : JsonSerializerContext
    {
    }
```

Since the `TypeGenerationSpec` is cached globally, it looks like the information generated for `Foo` in the two contexts will be the same, when in reality they should be different in each one.

Reply:
> Thinking more about this, I think we have a bug here:

Each `ContextGenerationSpec` has it's own [list](https://github.com/dotnet/runtime/blob/fc4a42746d9cf65ad3db4074207069ab9c168bc9/src/libraries/System.Text.Json/gen/ContextGenerationSpec.cs#L20) of `TypeGenerationSpec` instances, [which are unique and not shared](https://github.com/dotnet/runtime/blob/fc4a42746d9cf65ad3db4074207069ab9c168bc9/src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs#L174) with other contexts. The tests that derive from `RealWorldContextTests` verify that the generation mode for each context/type aligns with the user-specified configuration.

Reply:
Ah, I didn't see the cache getting cleared.

Back to the original comment, I still don't think calling `FastPathIsSupported()` is necessary here, if we just set the `generationMode` on the `typeGenerationSpec`, it shouldn't be necessary.

Reply:
> Back to the original comment, I still don't think calling FastPathIsSupported() is necessary here, if we just set the generationMode on the typeGenerationSpec, it shouldn't be necessary.

> The `generationMode` is getting flown into `TypeGenerationSpec.Initialize`. So the TypeGenerationSpec should have the generationMode set on it.

This only happens the first time the type is encountered. If a type occurs in another type's object graph (e.g as the type of a property or field) before we see a `[JsonSerializable]` for the type, then it will be initialized with either the context-wide mode or the type-specific mode for the "parent" class. When we eventually see a `[JsonSerializable]` for said type, we'll [exit early](https://github.com/dotnet/runtime/blob/93a1da7c391e178c0c1906dd724c0babf717a00e/src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs#L491-L494) from the `GetOrAddTypeGenerationSpec` method. That's when the [logic](https://github.com/dotnet/runtime/blob/93a1da7c391e178c0c1906dd724c0babf717a00e/src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs#L347-L370) to set/override the type-specific `TypeInfoPropertyName` and `GenerationMode` kicks in.

Reply:
> if we just set the generationMode on the typeGenerationSpec, it shouldn't be necessary.

👍  per offline discussion, I understand the suggestion better now and I've addressed it in the latest commit. Thanks!